File: EPs/EPs_cpp/ep_01.cpp
#include <iostream>
#include <cmath>
using namespace std;

/* Indica se o drone esta armado*/
bool armar(string drone, bool armado, bool decolado);
/**/
bool decolar(string drone, bool armado, bool decolado, double &altura);
/**/
void mudar_posicao(string drone, bool decolado, double &posicao, double &altura);
/**/
void pousar(string drone, bool &decolado, bool &armado, double &altura);

int main()
{
  bool armado, decolado;
  double altura = 0, posicao = 0;
  string drone;
  int escolha = 0;

  cout << "Simulador de Drones Skyrats" << endl;
  cout << "Qual drone voce esta usando?" << endl;
  cin >> drone;

  cout << "Posicao inicial: ";
  cin >> posicao;
  cout << "Altura inicial: ";
  cin >> altura;

  while (escolha != 5)
  {
    cout << "O que voce quer fazer com o drone?" << endl;
    cout << "\t1) Armar o drone" << endl;
    cout << "\t2) Decolar o drone" << endl;
    cout << "\t3) Mudar a posicao do drone" << endl;
    cout << "\t4) Pousar o drone" << endl;
    cout << "\t5) Sair" << endl;
    cin >> escolha;
    if (escolha == 1)
    {
      armado = armar(drone, armado, decolado);
    }
    else if (escolha == 2)
    {
      decolado = decolar(drone, armado, decolado, altura);
    }
    else if (escolha == 3)
    {
      mudar_posicao(drone, decolado, posicao, altura);
    }
    else if (escolha == 4)
    {
      pousar(drone, decolado, armado, altura);
    }
    else
    {
      return 0;
    }
  }
}

bool armar(string drone, bool armado, bool decolado)
{
  if (armado == true)
  {
    cout << drone << " ja foi armado" << endl;
    return true;
  }
  else
  {
    cout << drone << " armado com sucesso!" << endl;
    armado = true;
  }
  return true;
}

bool decolar(string drone, bool armado, bool decolado, double &altura)
{
  if (armado == false)
  {
    cout << drone << " precisa ser armado antes de decolar" << endl;
    return false;
  }
  if (decolado == true)
  {
    cout << drone << " ja esta voando" << endl;
    return true;
  }
  if (armado == true && altura == 0)
  {
    cout << "Para qual altura voce quer que seu drone decole?" << endl;
    cin >> altura;
    cout << drone << " decolado com sucesso para altura " << altura << endl;
    return true;
  }
}

void mudar_posicao(string drone, bool decolado, double &posicao, double &altura)
{
  double altura_inicial = altura;
  double posicao_inicial = posicao;

  double altura_t = altura_inicial;
  double posicao_t = posicao_inicial;

  double velocidade = 0, vel_x, vel_y, alpha;

  int tempo = 0;

  if (decolado == false)
  {
    cout << drone << " precisa ser decolado antes de mudar de posicao" << endl;
    return;
  }
  else
  {
    cout << "Posicao atual: "
         << "(" << posicao << "," << altura << ")" << endl;
    cout << "Qual posicao voce quer ir?" << endl;
    cout << "Posicao desejada: ";
    cin >> posicao;
    cout << "Altura desejada: ";
    cin >> altura;
    cout << "Velocidade desejada: ";
    cin >> velocidade;
    alpha = atan2(altura - altura_inicial, posicao - posicao_inicial);
    vel_x = velocidade * cos(alpha);
    vel_y = velocidade * sin(alpha);
  }

  while ((altura_t < altura) || (posicao_t < posicao))
  {
    cout << "Posicao no instante t" << tempo << ": (" << posicao_t << "," << altura_t << ")" << endl;
    posicao_t += vel_x;
    altura_t += vel_y;
    tempo++;
  }
  cout << drone << " foi para (" << posicao << "," << altura << ") com sucesso!" << endl;
}

void pousar(string drone, bool &decolado, bool &armado, double &altura)
{
  if (altura == 0)
  {
    cout << drone << " ja esta no chao" << endl;
    return;
  }
  else if (decolado == true)
  {
    cout << drone << " pousado com sucesso!" << endl;
    altura = 0;
    decolado = false;
  }
}

File: EPs/EPs_python/ep_01.py
# Simulação simplificada da estação de controle da Skyrats
import numpy as np


def main():
    print("Simulador de Drones Skyrats")

    drone = input("Qual drone voce esta usando?\n")

    global posicao
    posicao = float(input("Posição inicial: "))

    global altura
    altura = float(input("Altura inicial: "))

    global armado
    armado = False

    global voando
    voando = False

    escolha = 0
    while escolha != 5:
        print("O que voce quer fazer com o drone?")
        print("\t1) Armar o drone")
        print("\t2) Decolar o drone")
        print("\t3) Mudar a posição do drone")
        print("\t4) Pousar o drone")
        print("\t5) Sair")
        escolha = int(input(""))
        if escolha == 1:
            armado = armar(drone, armado, voando)
        elif escolha == 2:
            voando = decolar(drone, armado, voando)
        elif escolha == 3:
            mudar_posicao(drone, armado, voando)
        elif escolha == 4:
            pousar(drone, armado, voando)
        else:
            return 0


def armar(drone, armado, voando):
    # Verificar se o drone foi armado
    if armado == True:
        print("%s já foi armado!" % (drone))
        return True
    else:
        print("%s armado com sucesso!" % (drone))
        armado = True
    return True


def decolar(drone, armado, voando):
    global altura

    # Verificar se o drone foi armado
    if armado == False:
        print("%s precisa ser armado antes de decolar!" % (drone))
        return False

    # Verificar se o drone ja esta voando
    if voando == True:
        print("%s ja esta voando" % (drone))

    # Verificar se o drone esta armado e no chao
    if armado == True and altura == 0:
        altura = float(
            input("Para qual altura voce quer que seu drone decole?\n"))
        print("%s decolado com sucesso para altura %.1f" % (drone, altura))
    return True


def mudar_posicao(drone, armado, voando):
    global posicao
    global altura
    altura_inicial = altura
    posicao_inicial = posicao
    altura_t = altura_inicial
    posicao_t = posicao_inicial
    tempo = 0

    # Verificar se o drone ja foi decolado
    if voando == False:
        print("%s precisa ser decolado antes de mudar de posicao" % (drone))
        return
    #
    else:
        print("Posicao atual: (%.1f,%.1f)" % (posicao, altura))
        print("Qual posicao voce quer ir?")
        posicao = float(input("Posicao desejada: "))
        altura = float(input("Altura desejada: "))
        velocidade = float(input("Velocidade do drone: "))

        # Calcular as componentes do vetor velocidade
        alpha = np.arctan2(altura - altura_inicial, posicao - posicao_inicial)
        vel_x = velocidade * np.cos(alpha)
        vel_y = velocidade * np.sin(alpha)

        # Indicar a posicao e a altura do drone a cada segundo

        while posicao_t < posicao or altura_t < altura:
            print("Posicao no instante t%d: (%f, %f)" %
                  (tempo, posicao_t, altura_t))
            posicao_t += vel_x
            altura_t += vel_y
            tempo += 1

        print(drone, "foi para (%.1f,%.1f) com sucesso!" % (posicao, altura))


def pousar(drone, armado, voando):
    global altura

    # Verificar se o drone esta no chao
    if altura == 0:
        print("%s ja esta no chao" % (drone))
    elif voando == True:
        print("%s pousado com sucesso!" % (drone))
        altura = 0
        voando = False
        armado = False


main()
